File: enipath-frontfinal/src/app/services/statistics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface StudentStatistics {
  studentId: number;
  studentName: string;
  studentEmail: string;
  niveau: number;
  groupe: string;
  totalBadges: number;
  score: number;
  attendance: number;
  moyenneEvaluations: number;
  status: string;
}

export interface GroupeStatistics {
  groupe: string;
  niveau: number;
  totalStudents: number;
  moyenneScore: number;
  moyenneBadges: number;
  averageAttendance: number;
  students?: StudentStatistics[];
}

@Injectable({
  providedIn: 'root'
})
export class StatisticsService {

  private apiUrl = 'http://localhost:8081/api/statistics';

  constructor(private http: HttpClient) { }

  /**
   * Récupère les statistiques d'un étudiant (pour les professeurs)
   */
  getStudentStatistics(studentId: number): Observable<StudentStatistics> {
    return this.http.get<StudentStatistics>(`${this.apiUrl}/student/${studentId}`);
  }

  /**
   * Récupère les statistiques agrégées d'un groupe (pour les chefs)
   */
  getGroupStatistics(groupe: string, niveau: number): Observable<GroupeStatistics> {
    return this.http.get<GroupeStatistics>(`${this.apiUrl}/groupe/${groupe}/${niveau}`);
  }

  /**
   * Récupère les statistiques d'un groupe avec détails des étudiants (pour les profs)
   */
  getGroupStatisticsDetailed(groupe: string, niveau: number): Observable<GroupeStatistics> {
    return this.http.get<GroupeStatistics>(`${this.apiUrl}/groupe/${groupe}/${niveau}/detailed`);
  }

  /**
   * Récupère les statistiques de tous les groupes du département (pour les chefs)
   */
  getDepartmentStatistics(): Observable<GroupeStatistics[]> {
    return this.http.get<GroupeStatistics[]>(`${this.apiUrl}/department`);
  }

  /**
   * Récupère les statistiques des groupes d'un enseignant (sans détails)
   */
  getTeacherGroupsStatistics(teacherId: number): Observable<GroupeStatistics[]> {
    return this.http.get<GroupeStatistics[]>(`${this.apiUrl}/teacher/${teacherId}`);
  }

  /**
   * Récupère les statistiques des étudiants d'un enseignant (avec détails)
   */
  getTeacherStudentsStatistics(teacherId: number): Observable<StudentStatistics[]> {
    return this.http.get<StudentStatistics[]>(`${this.apiUrl}/teacher/${teacherId}/students`);
  }
}

File: enipathf/src/main/java/com/example/enipath/service/StatisticsService.java
package com.example.enipath.service;

import com.example.enipath.dto.GroupeStatisticsDto;
import com.example.enipath.dto.StudentStatisticsDto;
import com.example.enipath.model.Users.Enseignant;
import com.example.enipath.model.Users.Etudiant;
import com.example.enipath.model.academic.Groupe;
import com.example.enipath.repository.EnseignantRepository;
import com.example.enipath.repository.EtudiantRepository;
import com.example.enipath.repository.GroupeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class StatisticsService {
    
    @Autowired
    private EtudiantRepository etudiantRepository;
    
    @Autowired
    private EnseignantRepository enseignantRepository;
    
    @Autowired
    private GroupeRepository groupeRepository;
    
    /**
     * Récupère les statistiques individuelles d'un étudiant
     */
    public StudentStatisticsDto getStudentStatistics(Long studentId) {
        Etudiant student = etudiantRepository.findById(studentId)
            .orElseThrow(() -> new RuntimeException("Étudiant non trouvé"));
        
        return StudentStatisticsDto.builder()
            .studentId(student.getId())
            .studentName(student.getNom() + " " + student.getPrenom())
            .studentEmail(student.getEmail())
            .niveau(student.getNiveau())
            .groupe(student.getGroupe())
            .totalBadges(student.getTotalBadges())
            .score(student.getScore())
            .attendance(calculateAttendance(student)) // À implémenter selon votre logique
            .moyenneEvaluations(calculateAverageEvaluation(student)) // À implémenter
            .status("ACTIF")
            .build();
    }
    
    /**
     * Récupère les statistiques d'un groupe (agrégées)
     */
    public GroupeStatisticsDto getGroupStatistics(Character groupe, Integer niveau) {
        List<Etudiant> students = etudiantRepository.findStudentsByGroupeAndNiveau(groupe, niveau);
        
        double moyenneScore = students.stream()
            .mapToInt(Etudiant::getScore)
            .average()
            .orElse(0.0);
        
        double moyenneBadges = students.stream()
            .mapToInt(Etudiant::getTotalBadges)
            .average()
            .orElse(0.0);
        
        int averageAttendance = students.stream()
            .mapToInt(this::calculateAttendance)
            .sum() / Math.max(students.size(), 1);
        
        List<StudentStatisticsDto> studentStats = students.stream()
            .map(this::convertToStudentStatistics)
            .collect(Collectors.toList());
        
        return GroupeStatisticsDto.builder()
            .groupe(groupe)
            .niveau(niveau)
            .totalStudents(students.size())
            .moyenneScore(moyenneScore)
            .moyenneBadges(moyenneBadges)
            .averageAttendance(averageAttendance)
            .students(studentStats)
            .build();
    }
    
    /**
     * Récupère toutes les statistiques des groupes d'un enseignant (sans détails des étudiants)
     */
    public List<GroupeStatisticsDto> getTeacherGroupsStatistics(Long teacherId) {
        Enseignant teacher = enseignantRepository.findById(teacherId)
            .orElseThrow(() -> new RuntimeException("Enseignant non trouvé"));
        
        return teacher.getGroupes().stream()
            .map(groupe -> getGroupStatisticsAggregated(groupe.getCode(), groupe.getNiveau()))
            .collect(Collectors.toList());
    }
    
    /**
     * Statistiques agrégées d'un groupe (sans détails des étudiants)
     */
    public GroupeStatisticsDto getGroupStatisticsAggregated(Character groupe, Integer niveau) {
        List<Etudiant> students = etudiantRepository.findStudentsByGroupeAndNiveau(groupe, niveau);
        
        double moyenneScore = students.stream()
            .mapToInt(Etudiant::getScore)
            .average()
            .orElse(0.0);
        
        double moyenneBadges = students.stream()
            .mapToInt(Etudiant::getTotalBadges)
            .average()
            .orElse(0.0);
        
        int averageAttendance = students.stream()
            .mapToInt(this::calculateAttendance)
            .sum() / Math.max(students.size(), 1);
        
        return GroupeStatisticsDto.builder()
            .groupe(groupe)
            .niveau(niveau)
            .totalStudents(students.size())
            .moyenneScore(moyenneScore)
            .moyenneBadges(moyenneBadges)
            .averageAttendance(averageAttendance)
            .students(null) // Pas de détails pour le chef de département
            .build();
    }
    
    /**
     * Récupère les statistiques de TOUS les groupes pour un chef de département
     */
    public List<GroupeStatisticsDto> getDepartmentGroupsStatistics() {
        return groupeRepository.findAll().stream()
            .map(groupe -> getGroupStatisticsAggregated(groupe.getCode(), groupe.getNiveau()))
            .collect(Collectors.toList());
    }
    
    /**
     * Statistiques des étudiants d'un enseignant (pour les professeurs)
     */
    public List<StudentStatisticsDto> getTeacherStudentsStatistics(Long teacherId) {
        Enseignant teacher = enseignantRepository.findById(teacherId)
            .orElseThrow(() -> new RuntimeException("Enseignant non trouvé"));
        
        // Récupérer tous les étudiants de l'enseignant
        List<Etudiant> students = etudiantRepository.findByEnseignants_Id(teacherId);
        
        return students.stream()
            .map(this::convertToStudentStatistics)
            .collect(Collectors.toList());
    }
    
    // === MÉTHODES UTILITAIRES ===
    
    private StudentStatisticsDto convertToStudentStatistics(Etudiant student) {
        return StudentStatisticsDto.builder()
            .studentId(student.getId())
            .studentName(student.getNom() + " " + student.getPrenom())
            .studentEmail(student.getEmail())
            .niveau(student.getNiveau())
            .groupe(student.getGroupe())
            .totalBadges(student.getTotalBadges())
            .score(student.getScore())
            .attendance(calculateAttendance(student))
            .moyenneEvaluations(calculateAverageEvaluation(student))
            .status("ACTIF")
            .build();
    }
    
    private int calculateAttendance(Etudiant student) {
        // À implémenter selon votre logique (recalculer à partir des données d'assiduité)
        return 85; // Valeur par défaut
    }
    
    private double calculateAverageEvaluation(Etudiant student) {
        // À implémenter selon votre logique d'évaluation
        return student.getScore() / 100.0 * 20; // Conversion sur 20
    }
}
